### PR TITLE
Support for library with different name than crate name

### LINKED ISSUE #1099: Support for library with different name than crate name
An extension I am writing is part of multiple projects that have the same prefix, but for different purposes. As such, the crate name and the output binary may not be the same, configured at Cargo.toml. We would get the below error when doing `cargo pgx run` for these kind of crates.

```
Error: 
   0: Could not get shared object file from Cargo output.

Location:
   /Users/azam/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-pgx-0.7.4/src/command/install.rs:384

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

   0: cargo_pgx::command::install::find_library_file
      at /Users/azam/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-pgx-0.7.4/src/command/install.rs:351
   1: cargo_pgx::command::install::install_extension with pg_version=13.10 profile=Dev test=false features=["pg13"]
      at /Users/azam/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-pgx-0.7.4/src/command/install.rs:94
   2: cargo_pgx::command::run::run with pg_version=13.10 profile=Dev
      at /Users/azam/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-pgx-0.7.4/src/command/run.rs:94
   3: cargo_pgx::command::run::execute
      at /Users/azam/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-pgx-0.7.4/src/command/run.rs:55

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

The problem can be reproduced using any of the examples on this repo [(e.g. for strings)](https://github.com/tcdi/pgx/blob/master/pgx-examples/strings/Cargo.toml), by overriding the library name on the manifest as follows:

from:
```toml
[package]
name = "strings"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["cdylib"]
```

to:
```toml
[package]
name = "strings-pgx"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["cdylib"]
name = "strings"
```

A crate can only have one library, but the library does not necessarily be the crate name. Cargo manifest doc [here](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#library) does not explicitly say that, but under `[lib]` we can specify a `name` property that overrides the default library name, which is the crate name.

Seems like the function that finds the library file [(here)](https://github.com/tcdi/pgx/blob/f7ca21c4bf2028b5a17e009f2557864537e8f1d4/cargo-pgx/src/command/install.rs#L352) based on crate name. The function already have a reference to the whole manifest, could you add support for overridden library name as well?

### PR DESCRIPTION
**Pull request by azam**
This resolves #1099

### CODE REVIEW

**Review by eeeebbbbrrrr — APPROVED** [review]
This looks good.

I feel like we ought to document this somewhere, but pgx' "documentation" is kinda disorganized.  We'll get this documented somewhere later.

**Comment by eeeebbbbrrrr** [comment]
Oh, one change, @azam:  could you also update the `.github/workflows/tests.yml` file to include the new example?  Unfortunately we hardcode those...

**Comment by eeeebbbbrrrr** [comment]
Also, looks like it needs a `cargo fmt` too.  Thanks!  That's the test current test failure.

**Comment by azam** [comment]
@eeeebbbbrrrr
Sure! Pushed the changes. Found the doc on lib.name, so I also added a comment on changed line referencing the doc.

**Comment by azam** [comment]
Got to put a hold on this, I am stuck at getting the lib name [here](https://github.com/tcdi/pgx/blob/181ca0820ec9901a3dd94aad50a5f703cc2232a3/pgx-tests/src/framework.rs#L604). Should we read the cargo the manifest to get it, since there is no env var set for lib name (there is for binaries though `CARGO_BIN_NAME`)?

**Comment by eeeebbbbrrrr** [comment]
> Got to put a hold on this, I am stuck at getting the lib name [here](https://github.com/tcdi/pgx/blob/181ca0820ec9901a3dd94aad50a5f703cc2232a3/pgx-tests/src/framework.rs#L604). Should we read the cargo manifest to get it, since there is no env var set for lib name (there is for binaries though `CARGO_BIN_NAME`)?

That's a sneaky line of code!

Yeah, so I think now it makes sense to centralize the manifest handling, and use it here.  I think what might be okay is to add a `PgxManifest` type (or whatever) to the `pgx-pg-config` crate, which both `cargo-pgx` and `pgx-tests` already have a dependency for.

I'd go for the absolutely minimal implementation of this new `PgxManifest` type -- basically just a constructor that can parse the file and a public function to get the crate name we'd want everyone to use.

What do you think?

**Comment by azam** [comment]
I could do that, thinking on the line of exporting these functions:

* `get_crate_name(cargo_toml_path)`
* `get_crate__version(cargo_toml_path)`
* `get_extension_name(cargo_toml_path)`

Should I do this?

**Comment by eeeebbbbrrrr** [comment]
Yeah, that seems fine.  Thanks!

**Comment by azam** [comment]
@eeeebbbbrrrr 
Pushed my changes. Hope nothing breaks.

**Comment by eeeebbbbrrrr** [comment]
Since we just renamed to `pgrx`, this PR needs to be rebased against the develop branch.  Would you mind doing that?  

And if there's any references to "pgx" in your changeset they'll now need to be "pgrx".  Sorry for the trouble!  Growing pains!

**Comment by azam** [comment]
@eeeebbbbrrrr 
Rebased my repo!

**Comment by eeeebbbbrrrr** [comment]
This looks good.  I'm letting CI run it now.  Thanks @azam!

**Comment by eeeebbbbrrrr** [comment]
I'm not sure if you're allowed to see the test runner output?  <https://github.com/tcdi/pgrx/actions/runs/4738921279/jobs/8425118139>

But basically the tests are failing with:

```
Error: 
   0: Could not get shared object file from Cargo output.

Location:
   cargo-pgrx/src/command/install.rs:380

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

   0: cargo_pgrx::command::install::find_library_file
      at cargo-pgrx/src/command/install.rs:351
   1: cargo_pgrx::command::install::install_extension with pg_version=13.10 profile=Dev test=true features=["pg13", "pg_test", "cshim"]
      at cargo-pgrx/src/command/install.rs:94
   2: cargo_pgrx::command::install::execute
      at cargo-pgrx/src/command/install.rs:51
```

Oh, and this is for the `pgrx-tests` crate.  Any ideas?  Have you tried running `cargo test --all --features pg13 --no-default-features` locally?

File: cargo-pgrx/src/command/install.rs [hunk 1]
```diff
@@ -353,17 +353,13 @@ pub(crate) fn find_library_file(
...
-        return Err(eyre!("Could not get crate name from manifest."));
-    };
+    let target_name = manifest.target_name()?;
 
     let mut library_file = None;
     for message in build_command_messages {
         match message {
             cargo_metadata::Message::CompilerArtifact(artifact) => {
-                if artifact.target.name != *crate_name {
+                if artifact.target.name != *target_name {
```
Comment (azam) [inline cargo-pgrx/src/command/install.rs:362]: `artifact.target.name` is filled with crate name as-is (ie `pgrx-tests`) or lib name, depending on whether lib name has a name override or not, but `cargo_manifest::Manifest` auto fills lib name with crate name with hyphen replaced with underscore (ie `pgrx_tests`). Added `target_name` on `PgrxManifestExt` to return a value matching what `cargo_metadata` expects

**Comment by azam** [comment]
@eeeebbbbrrrr 
Apologize for the failure; borked my cargo-pgrx after the rebase, it kept the old version.
Fixed the error, re-ran the test locally and passed.
[test_out.log](https://github.com/tcdi/pgrx/files/11289857/test_out.log)

**Comment by eeeebbbbrrrr** [comment]
No need to apologize!  It's a complex system and that's why we have CI.  We appreciate your efforts on this PR!

**Review by eeeebbbbrrrr — APPROVED** [review]

**Pull request merged**